_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+
+  // Mocks
+  @MockBean
+  private CityRepository cityRepository;
+
+  @BeforeEach
+  void setUp() {
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+}
+```
+
+But caution: `@MockBean` is not an annotation from Mockito but from Spring!
+In the startup process Spring places the mock in the context, so that we don't need to do it ourselves.
+Wherever a bean claims to have its dependency satisfied, Spring injects the mock instead of the real object.
+This become handy if we want to have the same mock in different places.
+
+See [Mocking with Mockito and Spring Boot](https://reflectoring.io/spring-boot-mock/#mocking-with-mockito-and-spring-boot)
+for a deep dive how to mock Beans in Spring Boot.
+
+## Mockito Best Practices
+
+Knowing how to create the mocks, let's have a look at what we should avoid. Otherwise, our test cases are confusing
+what decreases our code's maintainability. It will save us much time debugging and doesn't let our team members guess
+what the intent of the test case is.
+
+### Avoid Concatenation in `setUp()`
+
+Even tough the test cases are reduced to a minimum, the readability suffers a lot. Besides, we must highly pay 
+attention to not break any other test cases. Like so, we avoid interfering other tests by overriding the setup.
+
+Avoid `setUp()` methods like this:
+
+```java
+  @BeforeEach
+  void setUp() {
+    expected = createCity();
+    cityRepository = Mockito.mock(CityRepository.class);
+    cityService = new CityServiceImpl(cityRepository);
+
+    // Avoid such complex declarations
+    Mockito.when(cityRepository.save(expected))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.find(expected.getId()))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.findByName(expected.getName()))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.findAllByCanton(expected.getCanton()))
+        .thenReturn(Collections.singleton(expected));
+    Mockito.when(cityRepository.findAllByCountry(expected.getCanton().getCountry()))
+        .thenReturn(Collections.singleton(expected));
+  }
+```
+
+To get simple test cases like this:
+
+```java
+  @Test
+  void save() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.save(expected));
+  }
+
+  @Test
+  void find() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.find(expected.getId()));
+  }
+
+  @Test
+  void delete() throws ElementNotFoundException {
+    cityService.delete(expected);
+    Mockito.verify(cityRepository).delete(expected);
+  }
+
+  @Test
+  void findByName() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.findByName(expected.getName()));
+  }
+```
+
+### Don't Recycle Mocks
+
+Having a complex scenario might entrap us to recycle our mocks. However, this will lead us sooner or later to unexpected
```

Comment:
Again, you're starting the section as if everybody knew what recycling mocks means.

Try this pattern:
1) intro sentence about recycling mocks
2) code example that's recycling mocks
3) explanation of the code example and why we shouldn't do it
4) code example how we should do it
5) explanation of the code example